---
title: 'Reproducible Research with RMarkdown'
author: "Marian L. Schmidt (marschmi@umich.edu; @micro_marian)"
date: "August 23, 2015"
output: ioslides_presentation  
---

## Who is Marian Schmidt?  
- I'm a native Michigander who likes to code in R.
- PhD Candidate in the Department of Ecology and Evolutionary Biology at the University of Michigan in [Vincent Denef's Lab](http://www-personal.umich.edu/~vdenef/).  
- My dissertation is focused on the spatial and temporal distributions and activities of freshwater bacteria.  
- I get to work in this beautiful place...
![](Figures/LakeMichigan.tif)

##  
When I'm not doing science-y things, you can find me here:  
![](Figures/climbing2.jpg) 

## Time to get to business  
![](Figures/Fig.1_alpha_SIGS_small.tif)

## As scientists, it should be our goal to perform **robust** and **reproducible** research.

- "**Robust research** is about doing small things that stack the deck in your favor to prevent mistakes." *~Vince Buffalo*
- **Reproducible research** may be repeated by other researchers with the same results. 

## Reproducibility can be difficult with genomic data.  
1. Genomics data is too large and high dimensional to easily inspect or visualize.  Usually, workflows involve multiple steps and it's not feasible to inspect every step.  
2. Unlike in the wet lab, we don't always know what to expect of our genomics data analysis.  
3. It's difficult to distinguish *good* from *bad* results.  
4. Scientific code is usually only run once to generate results for a publication, and is therefore more likely to contain (silent) bugs.  
    + **Silent errors** arise from code that may produce unknowingly incorrect output (rather than stop with an error message).


## What are the ingredients to robust and reproducible research?  

- **Work must be well documented!  Methods, code, and data must be made available to others!**  
- Adopt a cautious attitude and *check everything*.    
    + Vince Buffalo's golden rule of bioinformatics:  "Never ever trust your tools (or data)"  
    + Remember, "garbage in, garbage out" - an analysis is only as good as the data going in.



## What's the benefit for **_you?_**  
Yeah, it takes a lot of effort to be robust and reproducible.  However, *it will make your life (and science) easier!*  

- Most likely, you will have to re-run your analysis more than once.  
- In the future, you or a collaborator may have to re-visit part of the project.  
- You can make modularized parts of the project into re-useable tools for the future.  
    + Let me re-phrase: **Take the time to develop fequently used scripts into tools.**  
        + Then have your lab mates or collaborators test them and try to break them.


## 5 Recommendations for Robust Research  
1. Write code for humans, write data for computers  
2. Make assertions and be loud, in code and in your methods  
3. Use existing libraries (packages) whenever possible  
4. Prevent catastrophe and help reproducibility by making your data **read-only**  
5. Let data prove that it's high quality


## 1a. Write code for humans

**Code readability is very important.**  Code should be readable and broken down into small chunks that may be re-used.  (Do not re-write code to do the same task over and over again.)  

+ Comment code  
+ Adopt a [style guide](https://github.com/google/styleguide) 
    
If your code is more readable, then:  

- Your project is more reproducible.  
- It's easier to find and correct bugs.  
- You will be your friend in the future when you revisit the code.

## 1b. Write data for computers

**Let your computer do the work for you**  
Format your data so its easily read by your computer, not by you or other humans.  

- Code written for people to read requires cleaning and tidying to be processed by a computer.  
- Name data files in a consistent way.  
    + Automating tasks will be easier, which will prevent you from making trivial mistakes.

## 2. Make assertions and be loud, in code and in your methods  
Add tests within your code to make sure your code is doing what it is supposed to do.  
For example, use the `testthat` package

## 3. Use existing libraries (packages) whenever possible  
Do not try to re-invent the wheel while your performing your data anaylsis.  Use functions that have already been written and tested for you - so you take some of the work-load off of your shoulders!

## 4. Prevent catastrophe and help reproducibility by making your data **read-only**
Read-only is important because:  

- Modifying data can corrupt your results.  Therefore, making it less reproducible.  
- It's easy to lose track of how you have changed a file when you modify it in place.  

## 5. Let data prove that it's high quality








## 6 Recommendations for reproducible research  
1. Encapsulate the full project into one directory that is supported with version control.  
2. Release your code and data.  
3. Document everything and use code as documentation!
4. Make figures, tables, and statistics the results of scripts.  
5. Write code that uses relative paths.  
6. Always Set your seed.  


## 1. Encapsulate the full project into one directory that is supported with version control.  
The [Reproducible-Science-Curriculum Github repo for Reproducible Research Project Initialization](https://github.com/Reproducible-Science-Curriculum/rr-init) is a great place to start a reproducible research project.  

## 2. Release your code and data
It is simple.  Without your code and data, your research is not reproducible.

## 3. Document everything!   
**Bottom line:  Adopt a computing notebook that is as good as a wet-lab notebook**.

To fully reproduce a study, each step of analysis must be described in much more detail than can be included in a publication.  Therefore, additional documentation is essetial for reproducibility.

Include a record of your steps, where files are, where the came from, and what they contain.  Include “session info” in your document, preferably at the bottom. Session info lists the version of R that you’re using plus all of the packages you’ve loaded. There’s a sessionInfo() that’s included with base R (in the utils package), but [Dr. Karl Broman](http://kbroman.org/knitr_knutshell/pages/reproducible.html) recommends using the `session_info()` from the devtools package as it provides more detailed information in a nicer layout.

In your computing notebook:  

- Document your methods and workflows  
- Document the origin of all data in your project directory  
- Document **when** and **how** you downloaded the data  
- Record **data** version info  
- Record **software** version info (see below)

All the above information could be stored in a *README* file 


## 4. Make figures, tables, and statistics the results of scripts.

Using `inline code` can make the creation of tables much easier if the data changes!

## 5. Write code that uses relative paths.

Do not rely on hard-coded absolute paths (i.e. /Users/marschmi/Data/seq-data.csv or even ~/Data/seq-data.csv). Relative paths (i.e. Data/seq-data.csv) or command line arguments are better alternatives.

## 6. Always Set your seed  

If there is any randomizations of data or simulations, use `set.seed()` in the first code chunk.

Dr. Karl Browman suggests to open R and type runif(1, 0, 10^8) and then paste the resulting large number into set.seed() in the first code chunk. If you do this, then the random aspects of your analysis should be repeated the same way exactly whenever it is “knit.”






## What is R Markdown?   
* RMarkdown is a variant of Markdown that has embedded R code chunks to be used with `knitr` to make it easy to create reproducible web-based reports.  
      + **Markdown:**  A system for writing simple, readable text that is easily converted to html. 
          + Allows you to write using an easy-to-read, easy-to-write plain text format.              
*  Rmd -> md -> html (docx, pdf)  
*  Can include both text and code to execute  
  
    
    
## Why R Markdown?
A convenient tool for reproducible and dynamic reports with R!       

- Execute code with `knitr`.   
- Easy to learn syntax.  
- Include LaTeX equations.  
- Don't need to worry about page breaks or figure placement.  
- Consolidate your code and write up into a single file:  
    + Slideshows, pdfs, html documents, word files  
- It's **so easy** to use with version control with Git!   

## Simple Workflow

![](Figures/rmd_workflow.jpeg)


## Installing Rmarkdown and knitr
`install.packages("knitr")`
`install.packages("rmarkdown")` 


## How to Open an Rmd File
![](Figures/new.jpeg)
![](Figures/newRMD.jpeg)



## Choose Output
![](Figures/YAML_header.jpeg)

**YAML:**  A set of key value pairs at the start of your file.  Begin and end the header with a line of three dashes (- - -)

***R Studio template writes the YAML header for you***  

output: html_document  
output: pdf_document  
output: word_document  
output: beamer_presentation (beamer slideshow - pdf)  
output: ioslides_presentation (ioslides presentation - html)  



## Markdown basics 
Markdown is a simple formatting language that is easy to use

- Create lists with `*` or `+` sign   
      + like this
      + and this
- Use one or two asterisk marks to provide emphasis such as `*`*italics*`*` and `**`**bold**`**`.  Can even include tables:    

First Header  | Second Header
------------- | -------------
Content Cell  | Content Cell
Content Cell  | Content Cell


## Markdown basics 
![](Figures/Rmd_QuickRef.jpeg)

## R Code Chunks 

```
Code blocks display with fixed-width font
```

```{r fig.width = 4, fig.height = 3, message = FALSE}
#quick summary
library(ggplot2)
min(diamonds$price)
mean(diamonds$price)
max(diamonds$price)
```


## R Code Chunks 

```{r example chunk, echo = TRUE, eval = TRUE}
x <- sample(1:100, 5)
x
```

![](Figures/chunk_example.jpeg)

+ You can name the code chunk.  

+ echo = TRUE:  The code **will** be displayed.   

+ eval = TRUE: Yes, execute the code.

## R Code Chunk Arguments
![](Figures/knitr_arguments.jpeg)

## R Code Chunks:  Displaying Plots 
![](Figures/figure.jpeg)

```{r figure, echo = FALSE, fig.width = 5, fig.height = 4, fig.align='center'}
#quick plot of ggplot2's diamonds dataset
qplot(carat, price, data = diamonds, color = cut)
```


## Global Chunk Options

You may want to use the same set of chunk options throughout a document and you don't want to retype those options in every chunk.  

**Global chunk options are for you!**

![](global.jpeg)



## Inline R Code  
- You can evaluate expressions inline by enclosing the expression within a single back-tick qualified with `r`.

Last night, I saw `r 5 + 2` shooting stars!  



## Rendering document
1.  Run `rmarkdown::render("<filepath>")`  
2.  Click the **knit HTML** button at the top of the RStudio scripts pane  

When you render, R will:  

- Execute each embedded code chunk and insert the results into your report.  

- Build a new version of your report in the output file type.  

- Open a preview of the output file in the viewer pane.  

-  Save the output file in your working directory.  













